metadata-io/src/test/java/com/linkedin/metadata/entity/EbeanEntityServiceTest.java

```diff
@@ -650,6 +660,62 @@ public void testIngestListUrns() throws Exception {
   }
 
+  @Test
+  public void testRetention() throws Exception {
+    Urn entityUrn = Urn.createFromString("urn:li:corpuser:test1");
+
+    SystemMetadata metadata1 = new SystemMetadata();
+    metadata1.setLastObserved(1625792689);
+    metadata1.setRunId("run-123");
+
+    String aspectName = PegasusUtils.getAspectNameFromSchema(new CorpUserInfo().schema());
+
+    // Ingest CorpUserInfo Aspect
+    CorpUserInfo writeAspect1 = createCorpUserInfo("email@test.com");
+    _entityService.ingestAspect(entityUrn, aspectName, writeAspect1, TEST_AUDIT_STAMP, metadata1);
+    CorpUserInfo writeAspect1a = createCorpUserInfo("email_a@test.com");
+    _entityService.ingestAspect(entityUrn, aspectName, writeAspect1a, TEST_AUDIT_STAMP, metadata1);
+    CorpUserInfo writeAspect1b = createCorpUserInfo("email_b@test.com");
+    _entityService.ingestAspect(entityUrn, aspectName, writeAspect1b, TEST_AUDIT_STAMP, metadata1);
+
+    String aspectName2 = PegasusUtils.getAspectNameFromSchema(new Status().schema());
+    // Ingest Status Aspect
+    Status writeAspect2 = new Status().setRemoved(true);
+    _entityService.ingestAspect(entityUrn, aspectName2, writeAspect2, TEST_AUDIT_STAMP, metadata1);
+    Status writeAspect2a = new Status().setRemoved(false);
+    _entityService.ingestAspect(entityUrn, aspectName2, writeAspect2a, TEST_AUDIT_STAMP, metadata1);
+    Status writeAspect2b = new Status().setRemoved(true);
+    _entityService.ingestAspect(entityUrn, aspectName2, writeAspect2b, TEST_AUDIT_STAMP, metadata1);
+
+    assertEquals(_entityService.getAspect(entityUrn, aspectName, 1), writeAspect1);
+    assertEquals(_entityService.getAspect(entityUrn, aspectName2, 1), writeAspect2);
+
+    _retentionService.setRetention(null, null, new DataHubRetentionInfo().setRetentionPolicies(new RetentionArray(
+        ImmutableList.of(new Retention().setVersion(new VersionBasedRetention().setMaxVersions(2))))));
+    _retentionService.setRetention("corpuser", "status", new DataHubRetentionInfo().setRetentionPolicies(
+        new RetentionArray(
+            ImmutableList.of(new Retention().setVersion(new VersionBasedRetention().setMaxVersions(4))))));
+
+    // Ingest CorpUserInfo Aspect again
+    CorpUserInfo writeAspect1c = createCorpUserInfo("email_c@test.com");
+    _entityService.ingestAspect(entityUrn, aspectName, writeAspect1c, TEST_AUDIT_STAMP, metadata1);
+    // Ingest Status Aspect again
+    Status writeAspect2c = new Status().setRemoved(false);
+    _entityService.ingestAspect(entityUrn, aspectName2, writeAspect2c, TEST_AUDIT_STAMP, metadata1);
+
+    assertNull(_entityService.getAspect(entityUrn, aspectName, 1));
+    assertEquals(_entityService.getAspect(entityUrn, aspectName2, 1), writeAspect2);
+
+    // Reset retention policies
+    _retentionService.setRetention(null, null, new DataHubRetentionInfo().setRetentionPolicies(new RetentionArray(
+        ImmutableList.of(new Retention().setVersion(new VersionBasedRetention().setMaxVersions(1))))));
```

Comment:
What happens if max version is 10?

Reply:
It will just not do anything. It doesn't revive deleted aspects!

Reply:
but if you ingest more after setting it to 10, it will ingest without deleting until it reaches 10 versions